File: recipes/java-recipes/spring-boot-recipe/src/main/java/com/trendyol/stove/examples/java/spring/infra/components/product/api/ProductController.java
package com.trendyol.stove.examples.java.spring.infra.components.product.api;

import com.trendyol.stove.examples.java.spring.application.product.command.ProductApplicationService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Mono;

@RestController
@RequestMapping("/products")
public class ProductController {
  private final ProductApplicationService productService;

  public ProductController(ProductApplicationService productService) {
    this.productService = productService;
  }

  @PostMapping
  public Mono<ResponseEntity<?>> createProduct(@RequestBody ProductCreateRequest request) {
    return productService
        .create(request.getName(), request.getPrice())
        .onErrorContinue((throwable, o) -> ResponseEntity.badRequest().build())
        .then(Mono.fromCallable(() -> ResponseEntity.ok().build()));
  }
}

File: examples/spring-standalone-example/src/main/kotlin/stove/spring/standalone/example/infrastructure/api/ProductController.kt
package stove.spring.standalone.example.infrastructure.api

import kotlinx.coroutines.reactive.awaitSingle
import kotlinx.coroutines.reactor.mono
import org.springframework.http.codec.multipart.FilePart
import org.springframework.web.bind.annotation.*
import stove.spring.standalone.example.application.handlers.*

@RestController
@RequestMapping("/api")
class ProductController(private val productCreator: ProductCreator) {
  @GetMapping("/index")
  suspend fun get(
    @RequestParam(required = false) keyword: String
  ): String {
    return "Hi from Stove framework with $keyword"
  }

  @PostMapping("/product/create")
  suspend fun createProduct(
    @RequestBody productCreateRequest: ProductCreateRequest
  ): String = productCreator.create(productCreateRequest)

  @PostMapping("/product/import")
  suspend fun importFile(
    @RequestPart(name = "name") name: String,
    @RequestPart(name = "file") file: FilePart
  ): String {
    val content = file.content()
      .flatMap { mono { it.asInputStream().readAllBytes() } }
      .awaitSingle()
      .let { String(it) }
    return "File ${file.filename()} is imported with $name and content: $content"
  }
}
